post: organizing your react code (#61)
cloudx-labs/posts: posts/nachosource/01.organizing-your-react-code.md

## Don’t let yourself be pushed (by the urgencies)

We all know that things must be delivered as fast as they can be: the time pressure exists. But keep in mind that today’s mistakes are costing you in the future.
Don’t give more importance to matching the deadline than to building a robust logic. I’m sure that once you’ve learnt how to make things done it’s easy for you to deliver something that works. But don’t forget that this logic could need to be reused in the future, or even change.
So make sure that you won’t need to rebuild a whole component every time a style changes or a prop is added.

## Cohesion and coupling

These two are key concepts that will go along with wherever you code. Since they’ve already been explained way better than I could do here and now, I’m only talking about the core concept.

Imagine that your web has a couple of pages that share a section component. Let’s also say this section has a button that gets some data from an API, formats it and stores it somewhere in your app state, but this whole process is different each case.

> It’s easy to see that we are going to reuse at least a part of our code, keeping in mind that our goal is to have as little code as possible (we’ll return to this later). The point is, which is going to be our criteria here? Is there a standard to conform when designing our reusable component?

Reviewer: ```suggestion
It’s easy to see that we are going to reuse at least a part of our code, keeping in mind that our goal is to have as little code as possible (we’ll return to this later). The point is, which is going to be our criteria here? Is there a standard to conform to when designing our reusable component?
```

  <div style={styles}>
    <p>{text}</p>
    <button onClick={handleClick}>Click Me!</button>
  </div>
);

const Header = () => {
  const headerStyles = {
    backgroundColor: "green",
    fontWeight: "bold",
    width: "200px",
  };
  const headerTitle = "MyHeader";

  return (
    <div>
      <p>{headerTitle}</p>
      <MyComponent
        styles={headerStyles}
        text="Footer"
        handleClick={() => {
          console.log("I'm a Header");
        }}
      />
    </div>
  );
};

const Footer = () => {
  const footerStyles = {
    backgroundColor: "green",
    fontWeight: "lighter",
    width: "500px",
  };
  const footerTitle = "MyFooter";

  return (
    <div>
      <MyComponent
        styles={footerStyles}
        text="Footer"
        handleClick={() => {
          console.log("I'm a Footer");
        }}
      />
      <p>{footerTitle}</p>
    </div>
  );
};

const ContainerComponent = () => (
  <div>
    <Footer />
    <Header />
  </div>
);

export default ContainerComponent;
```

> Have you heard about the [_single-responsibility_](https://www.geeksforgeeks.org/how-to-use-single-responsibility-principle-in-reactjs/) principle?

## Organize your files

> One of the keys to have your code clean is to do the most with the least amount of code. When we get our components reusable we are also making an effort to improve the code’s readability and comprehension, compressing our files as possible.

Author: `compacting` could be used instead of `compressing`

Even if you’re not planning to reuse a code block all around the code, it’s nice to have it as an independent and self-descriptive component in its own file. This way it’s easy to get what it is used for (and, again, to abstract it from the context).

> If you do so, you are avoiding those thousand-lines files that make you read line by line of logic that you don’t need until you find your spot.

Author: Do you think `thousand-lines` is fine, or should I remove the `s`? 🤔

> If you do so, you are avoiding those thousand-lines files that make you read line by line of logic that you don’t need until you find your spot.

Reviewer: ```suggestion
If you do so, you are avoiding those thousand-line files that make you read line by line of logic that you don’t need until you find your spot.
```

One more thing about your files: once you’ve spread your blocks into functional components you can see that it’s hard to find each one amongst a large list of them, and even to name them!
Don’t forget to create some directories that gather these files using any criteria you want.

See the difference:

![files](./files.png)

> But what are the `index` files for? There you can organize the exportation in order to have your files clean when importing them.

Reviewer: ```suggestion
But what are the `index` files for? There you can organize the exports in order to have your files clean when importing them.
```

---
title: Organizing your React code: Cohesion and Coupling published: false description:
tags: 'react, javascript' cover_image:
---

Have you ever needed to spend a considerable amount of time reviewing your own code (written just a couple of months
ago) or been assigned to a new project just to find a pile of messy and eternal files?

I'm sure we've all had the feeling that a fix that took a day could've been solved in a fraction of that time by only
having the code a little more organized.

Consequences of spaghetti code are usually spending days instead of hours solving problems, code repetition,
inconsistencies between behaviors that should work the same way, inaccurate estimations and, of course, some headaches.

In this series I’ll try to share some concepts that help me when it comes to designing clean and maintainable
components, architecture and logic.

> But first, check this out: the person on the left is a responsible developer who follows the best practices and makes an

Author: This could also be in the ending

effort to improve his code's legibility when he has the chance. The one on the right is his twin brother, who thinks
it’s fine to _copy&paste_ the same logic over and over again and concentrate his code in a few files. Choose your cards
wisely!

![people](./assets/01.organizing-people.png)

## Cohesion and coupling

> These two are key concepts that will go along with wherever you code. Since they’ve already been explained way better

Reviewer 2: Shouldn't it be: "These two are key concepts that will go along with you wherever you code. Since they’ve already been explained way better"?

standard to conform to when designing our reusable component? The golden rule is: _the greatest cohesion, the lowest
coupling_.

What does this mean? Said in a few words, _cohesion_ is related with the criteria that gathers the inside elements of
your component, while _coupling_ has to do with the external relations established between your component and the
environment in which it is used (I mean, everything else). And let me just add that, as we have a components tree, these
concepts are to be transported to wherever you want, from less to more abstract.

> For more info about this topic [_have a read_](https://betterprogramming.pub/coupling-cohesion-552b022492b2).

### Why do we want our components to be _cohesive_ and _decoupled_?

The stronger relation you achieve between your inside elements, the more definite and accurate the identity of your
component is. The more abstracted you keep from the external requirements, the freer you get to reuse and adapt your
components to different situations.

Should I declare every single behavior the button might have when handling the click? Seems we could get a handler from

> our `props`, _decoupling_ our logic from the one depending on foreign elements. Are you sure it’s clean to fill your

Reviewer 3: ```suggestion
our `props`, _decoupling_ our logic from the one depending on foreign elements. Are you sure it's clean to fill your
```

> component with `if`s to validate what to render or not according to every possible use case? Let’s have a simple one-job

Reviewer 3: ```suggestion
component with `if`s to validate what to render or not according to every possible use case? Let's have a simple one-job
```

> component as external logic is out of our scope -since it is inside another component’s one.

Reviewer 3: ```suggestion
component as external logic is out of our scope -since it is inside another component's one.
```

> > Have you heard about the [_single-responsibility_](https://www.geeksforgeeks.org/how-to-use-single-responsibility-principle-in-reactjs/)
> > principle?

Reviewer 3: ```suggestion
> Have you heard about the [_single-responsibility_](https://www.geeksforgeeks.org/how-to-use-single-responsibility-principle-in-reactjs/) principle?
```

Let me share a simple example so we can figure this out:

```javascript
import React from "react";

const MyComponent = (props) => {
    const {isHeader = false, isFooter = false, text} = props;

    const headerTitle = isHeader ? "MyHeader" : "";
    const footerTitle = isFooter ? "MyFooter" : "";
    const styles = isHeader
        ? {
            backgroundColor: "red",
            fontWeight: "bold",
            width: "200px",
        }
        : {
            backgroundColor: "green",
            fontWeight: "lighter",
            width: "500px",
        };

>     const headerClickHandler = () => console.log("Getting header's data from API.");

Reviewer 3: ```suggestion
    const handleClickHeader = () => console.log("Getting header's data from API.");
```

>     const footerClickHandler = () => console.log("Getting footer's data from API.");

Reviewer 3: ```suggestion
    const handleClickFooter = () => console.log("Getting footer's data from API.");
```

            <p>{text}</p>
            <button onClick={isHeader ? headerClickHandler : footerClickHandler}>
                Click Me!
            </button>
            {isFooter && <p>{footerTitle}</p>}
        </div>
    );
};

export default MyComponent;
```

As we said, in these two scenarios the components' behaviour is similar. We can find differences in the `button`'s
*callback*, in the `labels`' content and in the subcomponents' order and styles, but the core idea of the component is
the same for both cases.

Besides, it's bothering to see that the component is full of validations that adjust what the component renders and does
according to the `props`' values. This means the component is coupled to the context's needs.

### Doing it better :smiley:

As all of this can be parameterized without losing the component's identity, we are going to extract the contextual
logic to achieve a simpler and more compact component that only does what it's meant to.

Following this same logic, we end up generating a few more components that gather their own specific logic (including a
parent `Container`):

```javascript
import React from "react";

> const MyComponent = ({styles, text, handleClick}) => (

Reviewer 3: ```suggestion
const MyComponent = ({styles, text, onClick}) => (
```

    <div style={styles}>
        <p>{text}</p>

>         <button onClick={handleClick}>Click Me!</button>

Reviewer 3: ```suggestion
        <button onClick={onClick}>Click Me!</button>
```

    </div>
);

const Header = () => {
    const headerStyles = {
        backgroundColor: "green",
        fontWeight: "bold",
        width: "200px",
    };
    const headerTitle = "MyHeader";

    return (
        <div>
            <p>{headerTitle}</p>
            <MyComponent
                styles={headerStyles}
                text="Footer"

>                 handleClick={() => {

Reviewer 3: ```suggestion
                onClick={() => {
```
